repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+    logger.info("{}:{} -- Obfuscated Quigley Matrix - Destroying Repose Routing Servlet", clusterId, nodeId)
+  }
+
+  override def configurationUpdated(configurationObject: SystemModel): Unit = {
+    //Set the current system model, and just update the nodes.
+    val interrogator = new SystemModelInterrogator(clusterId, nodeId)
+    localCluster.set(Option(interrogator.getLocalCluster(configurationObject).orElse(null)))
+    localNode.set(Option(interrogator.getLocalNode(configurationObject).orElse(null)))
+    defaultDestination.set(Option(interrogator.getDefaultDestination(configurationObject).orElse(null)))
+    destinations.set(localCluster.get().map(_.getDestinations) match {
+      case Some(dests) => getDestinations(dests.getEndpoint.asScala.toList) ++ getDestinations(dests.getTarget.asScala.toList)
+      case None => Map.empty[String, Destination]
+    })
+    domains.set(Option(new Clusters(configurationObject)))
+
+    if (logger.underlying.isDebugEnabled) {
+      localCluster.get().foreach { cluster =>
+        val clusterId: String = cluster.getId
+        //Build a list of the nodes in this cluster, just so we know what we're doing
+        val clusterNodes = cluster.getNodes.getNode.asScala.map(node => s"${node.getId}-${node.getHostname}")
+        logger.debug("{}:{} - Nodes for this router: {}", clusterId, nodeId, clusterNodes)
+        val destStrings = destinations.get()
+          .filter { case (_, dest) => dest.isInstanceOf[DestinationEndpoint] }
+          .map { case (_, endpoint) => s"${endpoint.getId}-${endpoint.asInstanceOf[DestinationEndpoint].getHostname}" }
+        logger.debug("{}:{} - Destinations for this router: {}", clusterId, nodeId, destStrings)
+      }
+    }
+    initialized = true
+
+    def getDestinations(destList: List[Destination]): Map[String, Destination] = {
+      destList.iterator.map(dest => dest.getId -> dest).toMap
+    }
+  }
+
+  override def isInitialized: Boolean = {
+    initialized
+  }
+
+  override def service(req: HttpServletRequest, resp: HttpServletResponse): Unit = {
+    if (!initialized) {
+      logger.error(NotInitializedMessage)
+      trySendError()
+    } else {
+      logger.trace("{} processing request...", this.getClass.getSimpleName)
+      Try(doRoute(new HttpServletRequestWrapper(req), resp)) recover {
+        case e =>
+          logger.error(FailedToRouteMessage, e)
+          trySendError()
+      }
+      logger.trace("{} returning response...", this.getClass.getSimpleName)
+    }
+
+    def trySendError(): Unit = {
+      if (!resp.isCommitted) {
+        resp.sendError(SC_INTERNAL_SERVER_ERROR, FailedToRouteMessage)
+      } else {
+        logger.error("Failed to Send Error due to response already being committed")
+      }
+    }
+  }
+
+  def doRoute(servletRequest: HttpServletRequestWrapper, servletResponse: HttpServletResponse): Unit = {
+    val attributeDestinations =
+      Option(servletRequest.getAttribute(CommonRequestAttributes.DESTINATIONS))
```

Comment:
We've talked about taking this out of being an attribute and making it a header like we do basically everything else. We should add a story to take a look at the viability of that. I see that the objects that make up the list have a uri, i wonder how necessary that is to it working?

Reply:
Created the story here:
* https://repose.atlassian.net/browse/REP-7174

Reply:
I would bet not very necessary. For the default destination, we set that value to the request URI. For other destinations, that value is set by whoever created the destination, which means that the request URI could potentially be ignored (and the destination URI used instead). That seems bad.

So yes, I think we should drop that `uri` field from the `RouteDestination`.

Reply:
We could always remove it during routing.

Reply:
That kind of solves my security and transparency concerns. You would still have to worry about forgery, but sure. At that point, though, why use a header?

Reply:
That is what the Header Normalization filter is for.

Reply:
So sure, maybe. That's why I called it out, after all. That solves the header coming in the door, and with some work, maybe even the header being added by another filter. I don't think there's ever a time where we take the response from an external service and copy a header onto the request, anyway. But my point is that I have to think about that. Why?

One reason that comes to mind is that we have logging around headers. Not so much around attributes. Of course, we could change that.

To me, per-request routing is functionality internal to Repose. There's no need to expose it to users, so why stick it in an HTTP field meant for transport? Handling it on the backend (as an attribute) works, and keeps me from thinking about the things I mentioned above (as much).